### ISSUE TITLE
Couldn't compile on Linux Manjaro 5.17

### ISSUE THREAD

**Issue by Nitrooo** [pkoutoupis/rapiddisk#77]
I swear I don't want to pester you, but I've just updated the kernel to:
`Linux pc 5.17.0-1-MANJARO #1 SMP PREEMPT Sun Jan 30 20:08:31 UTC 2022 x86_64 GNU/Linux`

And:

```
[user@pc rapiddisk]$ make
make -C src 
make[1]: Entering directory '/home/user/rapiddisk/src'
cc -c json.c -o json.o
cc -c rdsk.c -o rdsk.o
cc -c nvmet.c -o nvmet.o
cc -c sys.c -o sys.o
cc main.c -o rapiddisk json.o nvmet.o rdsk.o sys.o -ljansson
cc -c json.c -DSERVER -o json-server.o
cc -c net.c -o net.o
cc -c sys.c -DSERVER -o sys-server.o
cc daemon.c -o rapiddiskd json-server.o net.o sys-server.o -lpthread -ljansson -lmicrohttpd
Make has completed.
make[1]: Leaving directory '/home/user/rapiddisk/src'
make -C conf 
make[1]: Entering directory '/home/user/rapiddisk/conf'
Nothing to do.
make[1]: Leaving directory '/home/user/rapiddisk/conf'
make -C doc 
make[1]: Entering directory '/home/user/rapiddisk/doc'
Nothing to do.
make[1]: Leaving directory '/home/user/rapiddisk/doc'
make -C module 
make[1]: Entering directory '/home/user/rapiddisk/module'
make -C /lib/modules/5.17.0-1-MANJARO/build M=/home/user/rapiddisk/module
make[2]: Entering directory '/usr/lib/modules/5.17.0-1-MANJARO/build'
  CC [M]  /home/user/rapiddisk/module/rapiddisk.o
/home/user/rapiddisk/module/rapiddisk.c: In function ‘attach_device’:
/home/user/rapiddisk/module/rapiddisk.c:807:24: error: ‘GENHD_FL_SUPPRESS_PARTITION_INFO’ undeclared (first use in this function)
  807 |         disk->flags |= GENHD_FL_SUPPRESS_PARTITION_INFO;
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/user/rapiddisk/module/rapiddisk.c:807:24: note: each undeclared identifier is reported only once for each function it appears in
make[3]: *** [scripts/Makefile.build:288: /home/user/rapiddisk/module/rapiddisk.o] Error 1
make[2]: *** [Makefile:1831: /home/user/rapiddisk/module] Error 2
make[2]: Leaving directory '/usr/lib/modules/5.17.0-1-MANJARO/build'
make[1]: *** [Makefile:37: all] Error 2
make[1]: Leaving directory '/home/user/rapiddisk/module'
make: *** [Makefile:26: module] Error 2
```

**Comment by pkoutoupis** [comment]
No worries. You aren't pestering me. Looks like a simple fix. I will try to work on this tomorrow.

**Comment by pkoutoupis** [comment]
Here is a diff that "should compile". I haven't tested it yet...
```
diff --git a/module/rapiddisk.c b/module/rapiddisk.c
index 601a70c..b7d74b8 100644
--- a/module/rapiddisk.c
+++ b/module/rapiddisk.c
@@ -804,7 +804,11 @@ static int attach_device(int size)
        disk->fops = &rdsk_fops;
        disk->private_data = rdsk;
        disk->queue = rdsk->rdsk_queue;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,17,0)
+       disk->flags |= GENHD_FL_NO_PART;
+#else
        disk->flags |= GENHD_FL_SUPPRESS_PARTITION_INFO;
+#endif
        sprintf(disk->disk_name, "rd%d", num);
        set_capacity(disk, size);
```

**Comment by Nitrooo** [comment]
For some reason the patch didn't apply, but I edited `rapiddisk.c` manually. It compiles, installs and runs, but this is what happens when I try to attach a new RAM disk:

```
$ sudo rapiddisk -a 64
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis

Attached device rd0 of size 64 Mbytes
Killed
$ sudo rapiddisk -l
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis

Unable to locate any RapidDisk devices.
```

If I relaunch the `rapiddisk -a 64` command, it  just hangs forever.
I don't know if it's related to this patched version as I didn't test the 5.16 patch on the 5.16 kernel, or any other previous version/kernel. My current kernel is 5.17. Modules are loaded. 
Is there a log I can look into? I can open a new issue if necessary.

**Comment by pkoutoupis** [comment]
We will just use this ticket. I installed Manjaro and found the bug. I narrowed it down to the line of code. I will try to fix it tomorrow. Thank you again for your help.

**Comment by pkoutoupis** [comment]
Can you give this branch a try?
https://github.com/pkoutoupis/rapiddisk/tree/defect/5.15-queue-alloc-fix

I tried it in my Manjaro VM and all looked good.

**Comment by Nitrooo** [comment]
Sure, I'm gonna try it later this evening. Thank you very much for your super fast response.

**Comment by Nitrooo** [comment]
I'm trying the `defect/5.15-queue-alloc-fix` branch:

```
$ df
Filesystem      Size  Used Avail Use% Mounted on
.
.
.
/dev/md0        1,4T   34M  1,4T   1% /mnt/raid
$ sudo umount /mnt/raid/
$ sudo umount /mnt/raid/
umount: /mnt/raid/: not mounted.
$ sudo rapiddisk -a 64
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis

Attached device rd0 of size 64 Mbytes
$ sudo rapiddisk -l
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis

List of RapidDisk device(s):

 RapidDisk Device 1: rd0	Size (KB): 65536

List of RapidDisk-Cache mapping(s):

  None

$ sudo rapiddisk -m rd0 -b /dev/md0
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis


```

Here it hangs and I have to press CTRL + c:

```
$ sudo rapiddisk -m rd0 -b /dev/md0
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis

^CError. Unable to create map. Please verify all input values are correct.

```
I run it again:

```
$ sudo rapiddisk -m rd0 -b /dev/md0
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis

device-mapper: create ioctl on rc-wt_md0  failed: Device or resource busy
Command failed.
Error. Unable to create map. Please verify all input values are correct.

```

This is the relevant part of dmesg shown when the cache mapping hangs:
```
[   77.827610] device-mapper: rapiddisk-cache: Allocate 128KB (8B per) mem for 16384-entry cache(capacity:64MB, associativity:512, block size:8 sectors(4KB))
[   77.830187] ------------[ cut here ]------------
[   77.830189] kernel BUG at mm/slub.c:379!
[   77.830193] invalid opcode: 0000 [#1] PREEMPT SMP NOPTI
[   77.830195] CPU: 8 PID: 313 Comm: kworker/u64:6 Tainted: P           OE     5.17.0-1-MANJARO #1 e9effc340de5969a0880d76d9b285ad2aabd2372
[   77.830198] Hardware name: Micro-Star International Co., Ltd. MS-7C35/MEG X570 UNIFY (MS-7C35), BIOS A.C1 01/14/2022
[   77.830199] Workqueue: kcached do_work [rapiddisk_cache]
[   77.830204] RIP: 0010:kmem_cache_free+0x3e1/0x440
[   77.830207] Code: fd ff ff 49 89 da e9 6e fd ff ff 4d 89 e9 41 b8 01 00 00 00 4c 89 d1 48 89 da 48 89 ee 4c 89 f7 e8 a4 ac ff ff e9 a7 fd ff ff <0f> 0b 45 31 d2 e9 45 fd ff ff 49 8b 47 08 f0 48 83 28 01 75 b7 49
[   77.830208] RSP: 0018:ffffb17f80c3bbc0 EFLAGS: 00010246
[   77.830210] RAX: ffff8d0544405400 RBX: ffff8d0544405400 RCX: ffff8d05444054c8
[   77.830211] RDX: 0000000000216c08 RSI: fffffc3840000000 RDI: 0000000000000000
[   77.830212] RBP: fffffc3844110100 R08: ffff8d0544405400 R09: 0000000000000000
[   77.830213] R10: ffff8d0544405400 R11: 0000000000000000 R12: fffffc3844110100
[   77.830213] R13: ffffffffc0799b22 R14: ffff8d0541a7e500 R15: ffff8d0569ba7480
[   77.830214] FS:  0000000000000000(0000) GS:ffff8d141ec00000(0000) knlGS:0000000000000000
[   77.830215] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   77.830216] CR2: 00007f3c7c9377f8 CR3: 00000001b5610000 CR4: 0000000000750ee0
[   77.830217] PKRU: 55555554
[   77.830218] Call Trace:
[   77.830219]  <TASK>
[   77.830221]  endio+0x22/0x50 [dm_mod 1f035ce845ad985a8f4b0d8629aa29c4fef82e68]
[   77.830228]  ? bio_endio+0xd6/0x160
[   77.830231]  __submit_bio+0xbd/0x1d0
[   77.830233]  submit_bio_noacct+0x100/0x2a0
[   77.830234]  submit_bio+0x82/0x110
[   77.830236]  dispatch_io+0x1de/0x450 [dm_mod 1f035ce845ad985a8f4b0d8629aa29c4fef82e68]
[   77.830241]  ? list_next_page+0x20/0x20 [dm_mod 1f035ce845ad985a8f4b0d8629aa29c4fef82e68]
[   77.830246]  ? dm_io_client_create+0xb0/0xb0 [dm_mod 1f035ce845ad985a8f4b0d8629aa29c4fef82e68]
[   77.830251]  dm_io+0x127/0x240 [dm_mod 1f035ce845ad985a8f4b0d8629aa29c4fef82e68]
[   77.830255]  ? rc_uncached_io_callback+0xd0/0xd0 [rapiddisk_cache 322f49e566178c6112fe925eb3eb6af6b484b59f]
[   77.830257]  ? list_next_page+0x20/0x20 [dm_mod 1f035ce845ad985a8f4b0d8629aa29c4fef82e68]
[   77.830262]  ? dm_io_client_create+0xb0/0xb0 [dm_mod 1f035ce845ad985a8f4b0d8629aa29c4fef82e68]
[   77.830267]  do_work+0x15f/0x1e0 [rapiddisk_cache 322f49e566178c6112fe925eb3eb6af6b484b59f]
[   77.830269]  ? rc_uncached_io_callback+0xd0/0xd0 [rapiddisk_cache 322f49e566178c6112fe925eb3eb6af6b484b59f]
[   77.830271]  process_one_work+0x1e2/0x3b0
[   77.830273]  worker_thread+0x50/0x3a0
[   77.830275]  ? rescuer_thread+0x380/0x380
[   77.830276]  kthread+0xd6/0x100
[   77.830278]  ? kthread_complete_and_exit+0x20/0x20
[   77.830280]  ret_from_fork+0x1f/0x30
[   77.830283]  </TASK>
[   77.830284] Modules linked in: rfcomm mousedev cmac algif_hash algif_skcipher af_alg bnep joydev iwlmvm xfs squashfs nct6775 qrtr loop hwmon_vid mac80211 raid0 ip6t_REJECT nf_reject_ipv6 vfat libarc4 xt_hl md_mod fat ip6t_rt iwlwifi ipt_REJECT nf_reject_ipv4 xt_LOG r8169 cfg80211 nf_log_syslog realtek xt_limit btusb xt_addrtype btrtl xt_tcpudp btbcm btintel xt_conntrack btmtk nf_conntrack bluetooth nf_defrag_ipv6 nf_defrag_ipv4 libcrc32c ip6table_filter ip6_tables ecdh_generic intel_rapl_msr usbhid rfkill iptable_filter rapiddisk_cache(OE) rapiddisk(OE) dm_writecache dm_mod nvidia_uvm(POE) intel_rapl_common nvidia_drm(POE) edac_mce_amd snd_hda_codec_realtek kvm_amd nvidia_modeset(POE) snd_hda_codec_generic ccp ledtrig_audio snd_hda_codec_hdmi kvm snd_hda_intel snd_intel_dspcfg snd_intel_sdw_acpi snd_hda_codec irqbypass wmi_bmof mxm_wmi crct10dif_pclmul crc32_pclmul nvidia(POE) snd_hda_core ghash_clmulni_intel snd_hwdep aesni_intel snd_pcm crypto_simd cryptd snd_timer rapl snd sp5100_tco
[   77.830316]  pcspkr i2c_piix4 tpm_crb soundcore tpm_tis zenpower(OE) tpm_tis_core tpm wmi rng_core pinctrl_amd mac_hid acpi_cpufreq fuse crypto_user ip_tables x_tables ext4 crc32c_generic crc16 mbcache jbd2 xhci_pci crc32c_intel xhci_pci_renesas
[   77.830326] ---[ end trace 0000000000000000 ]---
[   77.830327] RIP: 0010:kmem_cache_free+0x3e1/0x440
[   77.830329] Code: fd ff ff 49 89 da e9 6e fd ff ff 4d 89 e9 41 b8 01 00 00 00 4c 89 d1 48 89 da 48 89 ee 4c 89 f7 e8 a4 ac ff ff e9 a7 fd ff ff <0f> 0b 45 31 d2 e9 45 fd ff ff 49 8b 47 08 f0 48 83 28 01 75 b7 49
[   77.830330] RSP: 0018:ffffb17f80c3bbc0 EFLAGS: 00010246
[   77.830331] RAX: ffff8d0544405400 RBX: ffff8d0544405400 RCX: ffff8d05444054c8
[   77.830332] RDX: 0000000000216c08 RSI: fffffc3840000000 RDI: 0000000000000000
[   77.830333] RBP: fffffc3844110100 R08: ffff8d0544405400 R09: 0000000000000000
[   77.830333] R10: ffff8d0544405400 R11: 0000000000000000 R12: fffffc3844110100
[   77.830334] R13: ffffffffc0799b22 R14: ffff8d0541a7e500 R15: ffff8d0569ba7480
[   77.830335] FS:  0000000000000000(0000) GS:ffff8d141ec00000(0000) knlGS:0000000000000000
[   77.830336] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   77.830337] CR2: 00007f3c7c9377f8 CR3: 00000001b5610000 CR4: 0000000000750ee0
[   77.830337] PKRU: 55555554
```

**Comment by pkoutoupis** [comment]
Interesting because I exercised both on Manjaro running a 5.16 kernel and it worked without issue. The 5.17 change is trivial. Do you see any warnings when compiling the modules?

**Comment by pkoutoupis** [comment]
Also, do you know which RC candidate you are using for 5.17? 1 or 2? I see that after 1 and before 2, a couple of dm I/O related patches went in but not sure if it entirely affects us here.

**Comment by Nitrooo** [comment]
No errors or warnings while compiling, just a bunch of `Deprecated feature: REMAKE_INITRD` while compiling the modules.

Creating a file system on the RAM drive:
```
$ sudo rapiddisk -a 1024
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis

Attached device rd0 of size 1024 Mbytes
$ sudo mkfs.ext4 /dev/rd0 
mke2fs 1.46.5 (30-Dec-2021)
Segmentation fault (core dumped)
```

**Comment by pkoutoupis** [comment]
OK. I reproduced it. Thanks again for your help. I do apologize for putting you through this hassle. You have been extremely helpful.

**Comment by pkoutoupis** [comment]
This patch will need to be placed on top of the development branch:
```
--- rapiddisk/module/rapiddisk.c.orig	2022-02-06 14:51:32.488620176 -0600
+++ rapiddisk/module/rapiddisk.c	2022-02-06 14:47:26.038800265 -0600
@@ -58,6 +58,7 @@
 
 /* ioctls */
 #define INVALID_CDQUERY_IOCTL	0x5331
+#define INVALID_CDQUERY_IOCTL2	0x5395
 #define RD_GET_STATS		0x0529
 
 static DEFINE_MUTEX(sysfs_mutex);
@@ -575,7 +576,11 @@ rdsk_make_request(struct request_queue *
 
 		err = rdsk_do_bvec(rdsk, bvec.bv_page, len,
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4,8,0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,16,0)
+				   bvec.bv_offset, bio_op(bio), sector);
+#else
 				   bvec.bv_offset, op_is_write(bio_op(bio)), sector);
+#endif
 #else
 				   bvec.bv_offset, rw, sector);
 #endif
@@ -603,13 +608,14 @@ out:
 	bio_endio(bio, err);
 #else
 	bio_endio(bio);
+#endif
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4,4,0)
 #if LINUX_VERSION_CODE < KERNEL_VERSION(5,16,0)
 	return BLK_QC_T_NONE;
-#endif
 #else
 	return;
 #endif
+#endif
 io_error:
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4,13,0)
 	bio->bi_status= err;
@@ -622,7 +628,6 @@ io_error:
 	return BLK_QC_T_NONE;
 #endif
 #endif
-#endif
 #if LINUX_VERSION_CODE < KERNEL_VERSION(3,2,0)
 	return 0;
 #endif
@@ -677,6 +682,7 @@ static int rdsk_ioctl(struct block_devic
 		mutex_unlock(&ioctl_mutex);
 		return error;
 	case INVALID_CDQUERY_IOCTL:
+	case INVALID_CDQUERY_IOCTL2:
 		return -EINVAL;
 	case RD_GET_STATS:
 		return copy_to_user((void __user *)arg,
```
Here is proof:
```
[petros@petros-virtualbox ~]$ cd rapiddisk/src/
[petros@petros-virtualbox src]$ make
cc -c json.c -o json.o
cc -c rdsk.c -o rdsk.o
cc -c nvmet.c -o nvmet.o
cc -c sys.c -o sys.o
cc main.c -o rapiddisk json.o nvmet.o rdsk.o sys.o -ljansson
cc -c json.c -DSERVER -o json-server.o
cc -c net.c -o net.o
cc -c sys.c -DSERVER -o sys-server.o
cc daemon.c -o rapiddiskd json-server.o net.o sys-server.o -lpthread -ljansson -lmicrohttpd
Make has completed.
[petros@petros-virtualbox src]$ sudo ./rapiddisk -a 128
[sudo] password for petros: 
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis

Attached device rd0 of size 128 Mbytes
[petros@petros-virtualbox src]$ uname -r
5.17.0-1-MANJARO
[petros@petros-virtualbox src]$ sudo mke2fs /dev/rd0
mke2fs 1.46.5 (30-Dec-2021)
Creating filesystem with 131072 1k blocks and 32768 inodes
Filesystem UUID: 5adfbaf6-a2d5-48e8-9931-27bd25be60de
Superblock backups stored on blocks: 
	8193, 24577, 40961, 57345, 73729

Allocating group tables: done                            
Writing inode tables: done                            
Writing superblocks and filesystem accounting information: done 

[petros@petros-virtualbox src]$ sudo mount /dev/rd0 /mnt/
[petros@petros-virtualbox src]$ touch /mnt/^C
[petros@petros-virtualbox src]$ df |grep rd0
/dev/rd0          121299       14    114732   1% /mnt
[petros@petros-virtualbox src]$ sudo touch /mnt/hello
[petros@petros-virtualbox src]$ ls /mnt/
hello  lost+found
```
And:
```
[petros@petros-virtualbox module]$ modinfo rapiddisk.ko|grep -e version -e vermagic
version:        8.0.1
srcversion:     614AB0B593CAC67ADC2BA48
vermagic:       5.17.0-1-MANJARO SMP preempt mod_unload 
```
I also realized that I was originally testing with 5.15 and not 5.16. For some reason I had thought that I was on 5.16. As you can see above I am now on 5.17 and you see proof (after the patch) of the driver working. I do apologize for putting you through this but you have been extremely helpful.

**Comment by pkoutoupis** [comment]
Also, before this exercise I have never used Manjaro and I am actually liking the distribution a lot. :-)

**Comment by Nitrooo** [comment]
It works now, apparently, but I will have to test it thoroughly. Thank you very much for your support and your great software.
One more question. When I map a RAM drive to a block device with `-p wt` or `-p wa`, it works just fine, but:

```
$ sudo rapiddisk -m rd0 -b /dev/md0 -p wb
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis

device-mapper: reload ioctl on rc-wb_md0 (253:0) failed: Invalid argument
Command failed.
Error. Unable to create map. Please verify all input values are correct.
```
And from dmesg:
```
[  669.959250] device-mapper: table: 253:0: writecache: Invalid magic in the superblock (-EINVAL)
[  669.959253] device-mapper: ioctl: error adding target to table
```
`dm-writecache` is loaded:
```
$ lsmod | grep writecache
dm_writecache          45056  0
dm_mod                184320  2 rapiddisk_cache,dm_writecache
```

**Comment by pkoutoupis** [comment]
I will need to work on that piece tomorrow. That is a user space utility issue because it relies on dm-writecache which I do not maintain. I wonder if they changed something in the driver code. Thanks again!

**Comment by Nitrooo** [comment]
Sure, take your time and thank you again.
Regarding Manjaro... yeah, I used Ubuntu-based distros for years, mainly Mint, then I switched to Manjaro and fell in love with the massive library of packages of the AUR repository as I have a lot of software installed from GitHub projects and I don't want to spend my time recompiling it or keeping track of the various PPAs. It's Arch, but for lazy people like me. It breaks from time to time, I've had issues with the Nvidia driver updates, but nothing catastrophic so far. :)

**Comment by pkoutoupis** [comment]
Hmmmm.....
In my Manjaro environment, it works.
```
[petros-virtualbox src]# sudo ./rapiddisk -a 128
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis

Attached device rd0 of size 128 Mbytes
[petros-virtualbox src]# sudo ./rapiddisk -m rd0 -b /dev/sdb -p wb
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis

Command to map rc-wb_sdb with rd0 and /dev/sdb has been sent.
Verify with "-l"

[petros-virtualbox src]# sudo ./rapiddisk -l
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis

List of RapidDisk device(s):

 RapidDisk Device 1: rd0	Size (KB): 131072

List of RapidDisk-Cache mapping(s):

 dm-writecache Target   1: rc-wb_sdb	Cache: rd0  Target: sdb (WRITEBACK)

[petros-virtualbox src]# uname -a
Linux petros-virtualbox 5.17.0-1-MANJARO #1 SMP PREEMPT Sun Jan 30 20:08:31 UTC 2022 x86_64 GNU/Linux
```
BUT it seems like dm_writecache is picky on what kind of devices it uses. On another machine I attempted to map to an NVMe and kept getting this:
```
[4948837.029285] device-mapper: table: 253:0: writecache: Origin data device lookup failed
[4948837.029314] device-mapper: ioctl: error adding target to table
```
And another SATA drive:
```
[4948707.813606] device-mapper: table: 253:0: len=78156289 not aligned to h/w logical block size 4096 of sdf
[4948707.813608] device-mapper: core: Cannot calculate initial queue limits
[4948707.813636] device-mapper: ioctl: unable to set up device queue for new table.
```
But then on the same machine, I used another SAS drive, no issues. No, I am not saying drive types matter but I think that certain drive parameters (such as logical/physical block sizes, etc.) make dm_writecache unhappy.

**Comment by Nitrooo** [comment]
To be honest I'm not familiar with that module or kernel development in general, for that matters.
In my case the device is a RAID 0 array created with mdadm. The sad thing is that I think my use case would benefit a lot from a writeback policy, since it's a very read/write intensive job and it doesn't need data persistence. I know it doesn't depend on you and you're already helping me greatly but, any idea?

**Comment by Nitrooo** [comment]
```
$ sudo dumpe2fs /dev/md0
dumpe2fs 1.46.5 (30-Dec-2021)
dumpe2fs: Bad magic number in super-block while trying to open /dev/md0
Couldn't find valid filesystem superblock.
/dev/md0 contains a xfs file system
```

Well, I wanted to see the magic number of the mdadm array, but even dumpe2fs complains about it not being valid.
As you said, dm_writecache must be picky, but apparently my very issues is related to the way I created the mdadm array. I found a thread about this issue and will look into it tomorrow. :)

**Comment by pkoutoupis** [comment]
OK. In the interim (probably tomorrow) I will push what I have to master. Thank you again!

**Comment by Nitrooo** [comment]
I managed to create an array with a valid super-block and the `-p wb` mapping succeeded!
Now, how do I interpret the output of `-s`?

```
$ sudo rapiddisk -s rc-wb_md0 
rapiddisk 8.0.1
Copyright 2011 - 2022 Petros Koutoupis

0 2414862336 writecache 0 8355967 0 0 342547 341572 11696 7 496 1885 9308 0 2 0
```

**Comment by pkoutoupis** [comment]
Awesome! Happy to hear. To answer your question: [here is the dm-writecache kernel doc](https://www.kernel.org/doc/html/latest/admin-guide/device-mapper/writecache.html)
```
Status:
1. error indicator - 0 if there was no error, otherwise error number
2. the number of blocks
3. the number of free blocks
4. the number of blocks under writeback
5. the number of read requests
6. the number of read requests that hit the cache
7. the number of write requests
8. the number of write requests that hit uncommitted block
9. the number of write requests that hit committed block
10. the number of write requests that bypass the cache
11. the number of write requests that are allocated in the cache
12. the number of write requests that are blocked on the freelist
13. the number of flush requests
14. the number of discard requests
```

**Comment by Nitrooo** [comment]
Thank you very much. :)

**Comment by pkoutoupis** [comment]
If it is ok, I would like to close this ticket as resolved (since I pushed the working code to master). The dev branch has been deleted, so moving forward just use master. Anyway, if you experience the same or similar issues, feel free to reopen this or a new ticket.
